doc/containers.md

```diff
@@ -379,3 +379,21 @@ used as a shortcut to set both soft and hard limit (e.g.
 `limits.kernel.nofile=3000`) to the same value. A resource with no explicitly
 configured limitation will be inherited from the process starting up the
 container. Note that this inheritance is not enforced by LXD but by the kernel.
+
+## Live migration
+LXD supports live migration of containers using [CRIU](http://criu.org). In
+order to optimize the memory transfer for a container LXD can be instructed to
+make use of CRIU's pre-copy features. This means LXD will request CRIU to
+perform a series of memory dumps for the container. After each dump LXD will
+send the memory dump to the specified remote. In an ideal scenario each memory
+dump will decrease the delta to the previous memory dump. When the delta is
+equal or smaller than the specified threshold LXD will request CRIU to perform
+a final memory dump and transfer it. If the threshold is not reached after
+a the maximum number of allowed iterations LXD will request a final memory dump
+from CRIU and migrate the container.
+
+Key                                      | Type    | Default | Live update | API extension        | Description
+:--                                      | :---    | :------ | :---------- | :------------        | :----------
+migration.pre\_copy.enabled              | boolean | false   | yes         | migration\_pre\_copy | Use CRIU's pre-copy feature for optimized container memory transfer.
+migration.pre\_copy.max                  | integer | 10      | yes         | migration\_pre\_copy | Maximum number of pre-copy iterations to be performed.
+migration.pre\_copy.pre\_migrated\_pages | integer | 70      | yes         | migration\_pre\_copy | Threshold (%) at which enough memory pages have been pre-copied.
```

Comment:
All configuration keys should be in the main table at the top.

I'd also recommend the following renames and changes to the descriptions:

 - migration.incremental.memory (was migration.pre_copy.enabled) => Incremental memory transfer of the container's memory to reduce downtime.
 - migration.incremental.memory.goal (was migration.pre_copy.pre_migrated_pages) => Percentage of memory to have in sync before stopping the container.
 - migration.incremental.memory.iterations (was migration.pre_copy.max) => Maximum number of transfer operations to go through before stopping the container.

This should be generic enough to apply to other live migration technologies (should we ever support something other than Linux/LXC/CRIU) and also let us plug in similar incremental transfer for other pieces of the container migration.

I think that also makes it more obvious to a user without CRIU knowledge as to what those knobs actually do.

Reply:
changed